p2panda-rs/src/storage_provider/traits/storage_provider.rs

```diff
@@ -0,0 +1,512 @@
+
+/// Trait which handles all high level storage queries and insertions.
+///
+/// This trait should be implemented on the root storage provider struct. It's definitions make up
+/// the the high level methods a p2panda client needs when interacting with data storage. It will
+/// be used for storing entries (`publish_entry`), getting required entry arguments when creating
+/// entries (`get_entry_args`) and retrieving a document id by entry hash (`get_document_by_entry`).
+/// Methods defined on `StorageEntry` and `StorageLog` for lower level access to their respective
+/// data structures will also be available.
+///
+/// The methods defined here are the minimum required for a working storage backend,
+/// additional custom methods can be added per implementation.
+///
+/// For example: if I wanted to use a SQLite backend, then I would first implement [`StorageLog`]
+/// and [`StorageEntry`] traits with all their required methods defined (they are required traits
+/// containing lower level accessors and setters for the respective data structures). With these
+/// traits defined [`StorageProvider`] is almost complete as it contains default definitions for
+/// most of it's methods (`get_entry_args` and `publish_entry` are defined below). The only one
+/// which needs defining is `get_document_by_entry`. It is also possible to over-ride the default
+/// definitions for any of the trait methods.
+#[async_trait]
+pub trait StorageProvider<StorageEntry: AsStorageEntry, StorageLog: AsStorageLog>:
+    EntryStore<StorageEntry> + LogStore<StorageLog>
+{
+    /// Params when making a request to get the next entry args for an author and document.
+    type EntryArgsRequest: AsEntryArgsRequest + Sync;
+
+    /// Response from a call to get next entry args for an author and document.
+    type EntryArgsResponse: AsEntryArgsResponse;
+
+    /// Params when making a request to publish a new entry.
+    type PublishEntryRequest: AsPublishEntryRequest + Sync;
+
+    /// Response from a call to publish a new entry.
+    type PublishEntryResponse: AsPublishEntryResponse;
+
+    /// Returns the related document for any entry.
+    ///
+    /// Every entry is part of a document and, through that, associated with a specific log id used
+    /// by this document and author. This method returns that document id by looking up the log
+    /// that the entry was stored in.
+    async fn get_document_by_entry(
+        &self,
+        entry_hash: &Hash,
+    ) -> Result<Option<DocumentId>, Box<dyn std::error::Error>>;
+
+    /// Returns required data (backlink and skiplink entry hashes, last sequence number and the
+    /// document's log_id) to encode a new bamboo entry.
+    async fn get_entry_args(
+        &self,
+        params: &Self::EntryArgsRequest,
+    ) -> Result<Self::EntryArgsResponse, Box<dyn std::error::Error>> {
+        // Validate the entry args request parameters.
+        params.validate().map_err(|error| format!("{:?}", error))?;
```

Comment:
So the errors returned are just strings?

Reply:
In this case, yeh.. it's a string wrapped in an `Error`, I had some troubles coercing the validation error types for unknown reasons.... I'll try and fix before merging. 

In general, this method can return any implementation defined error type as long as they implement the `Error` trait.

Reply:
Ok, if we add some more constraints to our existing `Validate` trait definition then we can correctly coerce the returned error type. Now it looks like this:

https://github.com/p2panda/p2panda/blob/372f2721eb6721abe21b05743197da00cd671d5c/p2panda-rs/src/lib.rs#L82-L89
`

Reply:
(`Error` and `'static` added)

Reply:
Not sure if adding the 'static lifetime is a bit of a hack... I'm never sure about that.

Reply:
Looks like the standard `Error` trait _needs_ to be `'static`: https://doc.rust-lang.org/std/error/trait.Error.html